### PR TITLE
Improve support for empty Git repository in Git view

### PR DESCRIPTION
**Pull request by westbury**
This fixes https://github.com/theia-ide/theia/issues/5435

There are two parts to this.  The first part changes 'log' function so that an empty array is returned when the Git repository is empty (no HEAD).  

Rather than check for an empty repository before every call to 'log', we check for the empty repository only after an error.  This was done to avoid an extra Git command on every 'log'.

Having made that change, a change to ScmAmendComponent was necessary to cause the 'last commit' section to appear when the initial commit is committed.

This can be tested by showing a newly-created empty repository in the Scm view.  Commit the initial commit and the 'last commit' section appears.  Remove the initial commit and it disappears.  You can remove the initial commit by deleting .git/refs/heads/master.

I recommend merging this PR before https://github.com/theia-ide/theia/pull/5451 .   I believe I can re-work that PR on top of this one to implement @svenefftinge's amending of the initial commit.

### CODE REVIEW

File: packages/git/src/node/dugite-git.ts [hunk 1]
```diff
@@ -657,13 +657,37 @@ export class DugiteGit implements Git {
             const file = Path.relative(this.getFsPath(repository), this.getFsPath(options.uri)) || '.';
             args.push(...[file]);
         }
-        const result = await this.exec(repository, args);
+
+        const successExitCodes = new Set([0, 128]);
+        let result = await this.exec(repository, args, { successExitCodes });
+        if (result.exitCode !== 0) {
+            // The 'log' command could potentially be valid when no HEAD if the revision range does not involve HEAD */
+            if ((!options || !options.range || options.range.toRevision === 'HEAD') && !await this.revParse(repository, { ref: 'HEAD' })) {
```
Comment (marechal-p) [inline packages/git/src/node/dugite-git.ts]: With the way you wrote it, if `options` is undefined, the `revParse` test will be executed, is this on purpose? Is git relying on HEAD when you don't specify anything?

**Comment by westbury** [comment]
> The Git History view displays a loading icon indefinitely, when it used to show an error message:

Thank you for testing this and finding this problem.  On investigating, the problem is caused firstly by the code in GitHistoryWidget addCommits() which sees no commits, ends up calling git.lsFiles and finds that there is a repository.  It does not set an error message because pathIsUnderVersionControl is true.  Then render() is called, it sees there are no commits, it also sees there is no error message, so it falls thru to the final 'else' and shows the progress indicator (assuming incorrectly that either some commits or an error will arrive at some point).

I'll sort this out tomorrow.

File: packages/git/src/node/dugite-git.ts [hunk 2]
```diff
@@ -657,13 +657,37 @@ export class DugiteGit implements Git {
...
                 .filter(item => item && item.length > 0));
     }
 
+    async revParse(repository: Repository, options: Git.Options.RevParse): Promise<string | undefined> {
+        const ref = options.ref;
+        const successExitCodes = new Set([0, 128]);
+        const result = await this.exec(repository, ['rev-parse', ref], { successExitCodes });
+        if (result.exitCode === 0) {
+            const sha = result.stdout;
+            return sha;
```
[the remark this answers is not in the document]
Comment (westbury) [reply packages/git/src/node/dugite-git.ts]: Changed as suggested

File: packages/git/src/node/dugite-git.ts (hunk 1 above)
Comment (westbury) [reply packages/git/src/node/dugite-git.ts]: Yes, you are correct that the default is 'HEAD' so I believe the code is correct.  I have refactored a little to make the code less confusing.

**Comment by westbury** [comment]
I have now fixed up the Git History view too.  Also changed the commit message to include 'Git History'.

The messages were a little confused, causing the 'There is no Git history available for in xxx' message.  Though not connected to this PR, I re-wrote the error message rendering to fix that.

**Comment by akosyakov** [comment]
@kittaakos would it be fine to merge it, since @marechal-p tested changes already?

**Comment by kittaakos** [comment]
> would it be fine to merge it

Apologies, I thought I have already approved this. I must have been confused with [this](https://github.com/theia-ide/theia/pull/5451#issuecomment-502312206).

Sure 👍 

One remark though, it was not necessary to expose `revParse` on the Git API as it is called from the backend only. But that's fine.

**Comment by westbury** [comment]
I'm fine for this to be merged.  As soon that this is merged I will rebase https://github.com/theia-ide/theia/pull/5451 and update it to allow modifying the initial commit.

**Comment by akosyakov** [comment]
@marechal-p Could you nominate @westbury as a Theia committer please? that he can merge as well

**Comment by marcdumais-work** [comment]
Note: @marechal-p is on vacation ATM, back middle of next week.

**Pull request merged**